### PR TITLE
Fixed non-determinate preprocessing on DataLoader

### PR DESCRIPTION
**Pull request by AlexanderRadionov**
Added ind_worker_queue parameter to data.DataLoader. It makes preprocessing determinate.

DataLoader in multiprocessing mode may cause non-deterministic issue. Even if `radom_seed` has frozen, each subprocess may get tasks in unstable order. This is caused by different I/O time while data loads. If you use augmentation while data loading, it makes results unreproduceble. Look at the https://discuss.pytorch.org/t/deterministic-non-deterministic-results-with-pytorch/9087

To fix this issue I have added the individual queue for each worker. In this case each worker get tasks in the stable order. In summary, subprocess produces the stable results.

To reproduce issue you may change ind_worker_queue to `False` and run the script several times.
Code to reproduce issue.
```
import time
import numpy as np
import random
from datetime import datetime
import torch
from torch.utils.data import DataLoader
from torch.utils.data import Dataset

GLOBAL_SEED = 1024

def set_seed(seed):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)

GLOBAL_WORKER_ID = None
def worker_init_fn(worker_id):
    global GLOBAL_WORKER_ID
    GLOBAL_WORKER_ID = worker_id
    set_seed(GLOBAL_SEED + worker_id)


class RandomAugmentationDataset(Dataset):
    def __init__(self, data):
        self.data = data

    def __getitem__(self, item):
        global GLOBAL_WORKER_ID
        rnd = datetime.now().microsecond % 5 # it should be unstable and not depend on random_seed
        if rnd < 3:
            time.sleep(rnd * 0.1)
        return item, GLOBAL_WORKER_ID, self.data[item] + np.random.random_sample(self.data[item].shape)

    def __len__(self):
        return self.data.shape[0]

set_seed(GLOBAL_SEED)
data = np.random.random_sample((128, 128))
dataset = RandomAugmentationDataset(data)

expected = [
    (2070.646908929146, [97, 0, 79, 2, 85, 121, 117, 56, 83, 43, 70, 60, 108, 47, 101, 95], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),
    (2063.337454820789, [71, 91, 20, 34, 107, 27, 52, 55, 69, 90, 127, 84, 81, 105, 46, 124], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]),
    (2031.0978096882036, [118, 86, 44, 14, 18, 92, 98, 36, 49, 13, 111, 28, 116, 10, 104, 87], [2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]),
    (2045.9375171891238, [103, 15, 19, 80, 29, 120, 1, 106, 123, 24, 39, 31, 3, 65, 45, 50], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),
    (2083.1401737400893, [11, 89, 12, 58, 54, 32, 67, 17, 110, 40, 82, 9, 30, 94, 125, 35], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]),
    (2035.8920455019213, [41, 74, 33, 68, 64, 8, 76, 42, 126, 53, 122, 26, 114, 75, 57, 112], [2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]),
    (2078.9789912514643, [109, 119, 113, 37, 5, 22, 59, 51, 62, 25, 38, 66, 6, 102, 73, 96], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),
    (2054.459126893385, [16, 72, 88, 115, 48, 61, 7, 93, 4, 77, 63, 100, 99, 21, 78, 23], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]),
]

s = time.time()
dataloader = DataLoader(dataset, batch_size=16, shuffle=True, num_workers=3, worker_init_fn=worker_init_fn, ind_worker_queue=True)
for i, (items, worker_ids, batch) in enumerate(dataloader):
    assert round(expected[i][0], 5) == round(np.sum(batch.numpy()), 5)
    assert expected[i][1] == items.numpy().tolist()

e = time.time()
print("time", e-s)
```

### CODE REVIEW

**Comment by SsnL** [comment]
Thanks! Could you add a test in `test_dataloader.py` as well?

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -193,7 +194,13 @@ def __init__(self, loader):
...
         if self.num_workers > 0:
             self.worker_init_fn = loader.worker_init_fn
-            self.index_queue = multiprocessing.SimpleQueue()
+            self.ind_worker_queue = loader.ind_worker_queue
+            if self.ind_worker_queue:
+                self.index_queue = [multiprocessing.SimpleQueue() for _ in range(self.num_workers)]
+                self.worker_queue_binding = itertools.cycle(range(self.num_workers))
+            else:
+                self.index_queue = [multiprocessing.SimpleQueue()]
+                self.worker_queue_binding = itertools.cycle([0])
```
Comment (meownoid) [inline torch/utils/data/dataloader.py]: `itertools.repeat(0)`

File: torch/utils/data/dataloader.py [hunk 2]
```diff
@@ -206,8 +213,9 @@ def __init__(self, loader):
             self.workers = [
                 multiprocessing.Process(
                     target=_worker_loop,
-                    args=(self.dataset, self.index_queue, self.worker_result_queue, self.collate_fn,
-                          base_seed + i, self.worker_init_fn, i))
+                    args=(self.dataset, self.index_queue[next(self.worker_queue_binding)],
```
Comment (meownoid) [inline torch/utils/data/dataloader.py]: Why not `self.index_queue[i]`?

**Comment by meownoid** [comment]
Personally I don't think that relying on iterator consumption order is a good idea. One accident `next` call can ruin everything.

**Comment by AlexanderRadionov** [comment]
@SsnL Done

File: torch/utils/data/dataloader.py [hunk 3]
```diff
@@ -282,7 +288,11 @@ def _put_indices(self):
         indices = next(self.sample_iter, None)
         if indices is None:
             return
-        self.index_queue.put((self.send_idx, indices))
+        self.index_queue[self.worker_queue_idx].put((self.send_idx, indices))
+        if self.ind_worker_queue:
+            self.worker_queue_idx += 1
```
Comment (SsnL) [inline torch/utils/data/dataloader.py]: `self.worker_queue_idx = (self.worker_queue_idx + 1) % self.num_workers` looks cleaner

File: torch/utils/data/dataloader.py [hunk 4]
```diff
@@ -353,6 +364,9 @@ class DataLoader(object):
         worker_init_fn (callable, optional): If not None, this will be called on each
             worker subprocess with the worker id as input, after seeding and before data
             loading. (default: None)
+        ind_worker_queue (bool, optional): If ``True``, the worker will binded to invididual
```
Comment (SsnL) [inline torch/utils/data/dataloader.py]: typo: individual

File: torch/utils/data/dataloader.py [hunk 5]
```diff
@@ -353,6 +364,9 @@ class DataLoader(object):
...
+            processing queue. (default: False)
+
```
Comment (SsnL) [inline torch/utils/data/dataloader.py]: Can you also add a `.. note:: ` on how to get deterministic results? Basically set ting`ind_worker_queue` and setting seed properly in `worker_init_fn`.

File: torch/utils/data/dataloader.py [hunk 6]
```diff
@@ -366,7 +380,7 @@ class DataLoader(object):
 
     def __init__(self, dataset, batch_size=1, shuffle=False, sampler=None, batch_sampler=None,
                  num_workers=0, collate_fn=default_collate, pin_memory=False, drop_last=False,
-                 timeout=0, worker_init_fn=None):
+                 timeout=0, worker_init_fn=None, ind_worker_queue=False):
```
Comment (SsnL) [inline torch/utils/data/dataloader.py]: Could you add a check that raise ValueError if this flag is set and `num_workers<=0`?

File: torch/utils/data/dataloader.py [hunk 7]
```diff
@@ -309,7 +319,8 @@ def _shutdown_workers(self):
             while not self.data_queue.empty():
                 self.data_queue.get()
             for _ in self.workers:
-                self.index_queue.put(None)
+                for q in self.index_queue:
```
Comment (SsnL) [inline torch/utils/data/dataloader.py]: This is wrong. When `ind_worker_queue` is False, you need to put `num_worker` Nones.

**Review by ssnl — COMMENTED** [review]
Sorry for the late review. Thanks for the contribution. I think after addressing the comments and resolve conflict, this will be good to go :).

File: torch/utils/data/dataloader.py (hunk 6 above)
Comment (AlexanderRadionov) [reply torch/utils/data/dataloader.py]: Done.

**Comment by SsnL** [comment]
Sorry I forgot to enable CI on this PR. Should be testing now.

File: torch/utils/data/dataloader.py (hunk 7 above)
Comment (AlexanderRadionov) [reply torch/utils/data/dataloader.py]: I think this is correct. The cycle by ``for _ in self.workers`` gurantee that ``index_queue`` get the ``num_worker`` Nones. I have changed ``index_queue`` type from Queue to list of Queues. Because of this, I put None to all ``q`` in ``self.index_queue``. In case when ``ind_worker_queue`` if False, the ``self.index_queeue`` will list with only one Queue insight.

File: torch/utils/data/dataloader.py (hunk 5 above)
Comment (AlexanderRadionov) [reply torch/utils/data/dataloader.py]: Done.

File: torch/utils/data/dataloader.py (hunk 7 above)
Comment (SsnL) [reply torch/utils/data/dataloader.py]: Yes, I am aware that you changed `index_queue` to a list of queues. My point was that if `ind_worker_queue=False`, the single queue in `index_queue` should receive `num_worker` `None`s, rather than just one `None`.

Your new change is also incorrect. The behavior should be, if `ind_worker_queue=False`, put  `num_worker` `None`s into the single queue. Otherwise, put a `None` in each queue.

By the way, could you change the name `index_queue` to `index_queue` to avoid confusion? Thanks!

Comment (AlexanderRadionov) [reply torch/utils/data/dataloader.py]: OK, thank you for attention. Fixed.

**Comment by apaszke** [comment]
@AlexanderRadionov do you have any benchmarks for the new data loader? Me and @colesbury are worried that it will lower the throughput, and it's important for training on e.g. 8x Volta GPUs.

**Comment by AlexanderRadionov** [comment]
I made some syntetic tests, just iterate by dataloader. And i have not see any issue.
Here are code:
```
import time
from datetime import datetime
import numpy as np
import random
import torch
from torch.utils.data import DataLoader
from torch.utils.data import Dataset

GLOBAL_SEED = 1024

def set_seed(seed):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)

GLOBAL_WORKER_ID = None
def worker_init_fn(worker_id):
    global GLOBAL_WORKER_ID
    GLOBAL_WORKER_ID = worker_id
    set_seed(GLOBAL_SEED + worker_id)


class RandomAugmentationDataset(Dataset):
    def __init__(self, data):
        self.data = data

    def __getitem__(self, item):
        global GLOBAL_WORKER_ID
        return item, GLOBAL_WORKER_ID, self.data[item] + np.random.random_sample(self.data[item].shape)

    def __len__(self):
        return self.data.shape[0]

set_seed(GLOBAL_SEED)
data = np.random.random_sample((12800, 12800))
dataset = RandomAugmentationDataset(data)

s = time.time()
for i in range(0, 11):
    dataloader = DataLoader(dataset, batch_size=16, shuffle=True, num_workers=3, worker_init_fn=worker_init_fn, ind_worker_queue=True)
    for i, (items, worker_ids, batch) in enumerate(dataloader):
        pass
e = time.time()
print("time", e-s)
```
I run it with `ind_worker_queue=True` and `ind_worker_queue=False` several times.
`False` (default):
time 17.274786710739136
time 16.919730186462402
`True` (overrided):
time 16.521615266799927
time 16.571028470993042

I think my changes should not cause slowdown because individual worker behavior is not default.

**Comment by apaszke** [comment]
I'm thinking more of a real Imagenet-like workload that actually has to read from disk, and some requests may get stalled for longer than others, slowing down the whole loading process (unlike previously).

**Comment by AlexanderRadionov** [comment]
I tried to emulate read time difference in my original code posted at head of merge request (random sleep). Here are results:

`False` (default):
time 123.02024149894714
time 123.15355587005615

`True` (overrided):
time 126.91952013969421
time 126.23874187469482

Do you think is it problem for non-default individual?

**Comment by SsnL** [comment]
@apaszke Yes the process will be slowed. But have reproducible results is very important. Do you have suggestions on how to do multiprocess deterministic dataloading in a faster way?

**Comment by apaszke** [comment]
@AlexanderRadionov what was the duration of the sleep you've used? I'd still want to try this on a real machine before merging, but I think it's ok. The only change I'd request would be to remove the flag, and make this mode always enabled.

**Comment by AlexanderRadionov** [comment]
I've used sleep random [0.001; 0.003] duration.

Also I tested on real dataset data. Dataset have 96265 images, total 8,7G.

`False` (default)
3 workers: 237s
6 workers: 178s

`True` (overrided)
3 workers: 242s
6 workers: 177s

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
LGTM. I just want to run an real-world ImageNet benchmark and should be good to merge

**Comment by AlexanderRadionov** [comment]
@apaszke  I have no enough resources to run ImageNet benchmark.

**Comment by apaszke** [comment]
@AlexanderRadionov no worries, I wasn't requesting that from you. I was going to do it myself

**Comment by grafi-tt** [comment]
Just FYI, another strategy to ensuring determinacy is storing PRNG states on shared memory. It doesn't requires fixing worker assignment, though its implementation is complex and serialization / deserialization overheads are incurred.

A PR to Chainer https://github.com/chainer/chainer/pull/4230 uses this strategy, because fixing process assignment is impossible in Chainer as it uses `multiprocessing.Pool`.

**Comment by apaszke** [comment]
@grafi-tt I'm not sure how this would work. You need to ensure that the sequence of locks for the PRNG will be the same at every run, but this is very hard or prohibitively slow if you sample multiple times while loading a data item (e.g. sample a mask, sample a rotation, ...).

**Comment by grafi-tt** [comment]
@apaszke It's simple: cache the state. When a worker process is resumed, a PRNG state is deserialized from shared memory (by e.g. `numpy.RandomState.set_state()`). When the worker process yields a result, the PRNG state is serialized to shared memory (by e.g. `numpy.RandomState.get_state()`). During the worker is running, there is no overhead.

**Comment by apaszke** [comment]
@grafi-tt oh I see. This makes sense but is still not ideal:
- you still need a bunch of synchronization, and everyone needs to lock the shared memory and put their random state in it, while everyone else is just waiting (this is a fairly expensive memcpy)
- you really want to make the random sequences in different workers to seem independent of each other. If I understand correctly you will can end up with this behavior with two workers:
    - worker 1 starts, reads state 0
    - worker 2 starts, reads state 0
    - both workers do some work and sample e.g. twice. since both started from the same random state, they both end up in state 1
    - worker 1 finishes, saves state 1
    - worker 2 finishes, saves state 1
    - then the process repeats, and the sampling results are identical in both workers.

**Comment by brando90** [comment]
is this on pytorch 3.1?

**Comment by SsnL** [comment]
no

**Comment by brando90** [comment]
@SsnL then were is this?

**Comment by SsnL** [comment]
It’s on master but not in a release yet.

**Comment by brando90** [comment]
@SsnL when will it be released. I'd be really nice to have determinism in the data loader.

**Comment by SsnL** [comment]
I am on mobile so i can’t link it. But you can search the issue list, there
is a tracking issue for 0.4 release. There are still some rough edges to
smooth out. Although I don’t know the exact time, but we are actively
working on it. That is why I suggested building from source on your forum
post.

It is not hard at all to modify the files as this PR by yourself either
since IIRC this PR is entirely in python.

**Comment by derek-kim** [comment]
@AlexanderRadionov @SsnL @apaszke I agree that having reproducible results is very important, but that doesn't mean that you have to synchronize workers. Multiprocessing are working non-deterministic in nature and if you try have some unnatural synchronization there must be some performance trade-off like even if it's trivial. Also the synchronization makes the code more complicated and harder to maintain.

I think what you really need is a deterministic data augmentation function, rather than a deterministic data loader.

```
set_seed(GLOBAL_SEED)

class SomeDataset(Datset):
    def __init__(self, data):
        self.data = data
        self.random_numbers = make_random_numbers_in_advance(len(self.data))

    def __getitem__(self, idx):
        random_number = self.random_numbers[idx]
        # do some works with random_number
        return processed_item
```
